### PR TITLE
feat: Ability to add custom scripts to Livechat bundle

### PR DESCRIPTION
**Pull request by cabaceira**
## Proposed changes (including videos or screenshots)

https://github.com/RocketChat/Rocket.Chat/assets/213379/12a7a439-92ef-4983-a193-51d7a02f9ccb

https://github.com/RocketChat/Rocket.Chat/assets/213379/a0d02955-0f81-4208-a810-6f4b181eb99a

An example showing the scripts and css being used together with the widget inside an iframe.

https://github.com/RocketChat/Rocket.Chat/assets/213379/5b65dd6c-7340-40e8-ba77-6e8d5d27f517

This is a sponsored feature by Amerisave that needs the ability to add/load custom scripts inside the live chat widget, which is embedded as an iFrame so that the parent window cannot interfere with the content of what is embedded in the iFrame.

Rocket.Chat already offers the ability to handle external/custom scripts in the web client, for example, the same approach is implemented here for Livechat.  It is possbile to include several js script libraries separated by a “,” . Those scripts will be loaded in the context of the livechat widget as shown on the sample video.

The same is valid for custom css classes. Those can be expressed in a list separated by “;” and will be included in the context of the doom where the livechat widget runs.

### CODE REVIEW

File: apps/meteor/ee/app/livechat-enterprise/server/settings.ts [hunk 1]
```diff
@@ -157,6 +157,32 @@ export const createSettings = async (): Promise<void> => {
...
 
+	await settingsRegistry.add('Livechat_AdditionalWidgetScripts', '', {
+		type: 'string',
+		group: 'Omnichannel',
+		section: 'Livechat',
+		enterprise: true,
+		invalidValue: "",
+		multiline: true,
+		i18nLabel: 'Livechat_AdditionalWidgetScripts',
+		i18nDescription: 'Additional_script_to_embed_in_the_livechat_widget',
```
Comment (MartinSchoeler) [inline apps/meteor/ee/app/livechat-enterprise/server/settings.ts:168]: We probably need more information on how to use the scripts, for example, it seems that scripts are separated by a comma, going by the code, maybe give an example of valid additional script

File: apps/meteor/ee/app/livechat-enterprise/server/settings.ts [hunk 2]
```diff
@@ -157,6 +157,32 @@ export const createSettings = async (): Promise<void> => {
 		});
 	});
 
+	await settingsRegistry.add('Livechat_AdditionalWidgetScripts', '', {
+		type: 'string',
```
Comment (MartinSchoeler) [inline apps/meteor/ee/app/livechat-enterprise/server/settings.ts:161]: We should probably use a `type: 'code', code: 'javascript'` in this case

File: apps/meteor/ee/app/livechat-enterprise/server/settings.ts [hunk 3]
```diff
@@ -157,6 +157,32 @@ export const createSettings = async (): Promise<void> => {
...
+
+	await settingsRegistry.add('Livechat_WidgetLayoutClasses', '', {
+		type: 'string',
+		group: 'Omnichannel',
+		section: 'Livechat',
+		enterprise: true,
+		invalidValue: "",
+		multiline: true,
+		i18nLabel: 'Livechat_WidgetLayoutClasses',
+		i18nDescription: 'Additional_classes_to_add_in_body_of_livechat_widget',
```
Comment (MartinSchoeler) [inline apps/meteor/ee/app/livechat-enterprise/server/settings.ts:181]: Same here, we should specify how the classes are being separated (with comma)

File: apps/meteor/app/livechat/server/livechat.ts [hunk 4]
```diff
@@ -18,24 +23,39 @@ WebApp.connectHandlers.use('/livechat', (req, res, next) => {
 	}
 
 	res.setHeader('content-type', 'text/html; charset=utf-8');
-
+	const liveChatAdditionalScripts = settings.get<string>('Livechat_AdditionalWidgetScripts');
```
Comment (MartinSchoeler) [inline apps/meteor/app/livechat/server/livechat.ts:26]: When I try to use an additional script I get this error

Screenshot 2023-12-13 at 14 26 28

Do we need a specific setting to use this (to embed livechat we usually have to change some CORS settings)

**Review by MartinSchoeler — CHANGES_REQUESTED** [review]

File: apps/meteor/ee/app/livechat-enterprise/server/settings.ts (hunk 2 above)
Comment (cabaceira) [reply apps/meteor/ee/app/livechat-enterprise/server/settings.ts:161]: Well, its a CSV list, more of a string than node no ?

File: apps/meteor/app/livechat/server/livechat.ts (hunk 4 above)
Comment (cabaceira) [reply apps/meteor/app/livechat/server/livechat.ts:26]: How were you trying to use the script ?  i want to verify that on my end.

**Review by MartinSchoeler — CHANGES_REQUESTED** [review]
Also it seems that older scripts are not being removed from the page

Screenshot 2023-12-13 at 15 22 14

File: apps/meteor/app/livechat/server/livechat.ts (hunk 4 above)
Comment (MartinSchoeler) [reply apps/meteor/app/livechat/server/livechat.ts:26]: I was trying to use javascript directly, now I see that there we should be adding just the src there, we should probably specify that the src property of the tag will be the one set by the setting

File: apps/meteor/ee/app/livechat-enterprise/server/settings.ts (hunk 2 above)
Comment (MartinSchoeler) [reply apps/meteor/ee/app/livechat-enterprise/server/settings.ts:161]: Indeed, now I've understood what the setting adds to the widget

File: apps/meteor/ee/app/livechat-enterprise/server/settings.ts (hunk 3 above)
Comment (cabaceira) [reply apps/meteor/ee/app/livechat-enterprise/server/settings.ts:181]: I've added a more detailed explanation on the setting description and provided examples of usage.

File: apps/meteor/ee/app/livechat-enterprise/server/settings.ts (hunk 1 above)
Comment (cabaceira) [reply apps/meteor/ee/app/livechat-enterprise/server/settings.ts:168]: I've added a more detailed explanation on the setting description and provided examples of usage.

File: apps/meteor/app/livechat/server/livechat.ts (hunk 4 above)
Comment (cabaceira) [reply apps/meteor/app/livechat/server/livechat.ts:26]: I verified using the javascript elements in the iframe inside the livechat widget.

image

Comment (cabaceira) [reply apps/meteor/app/livechat/server/livechat.ts]: https://github.com/RocketChat/Rocket.Chat/assets/213379/c9a6bd61-42ff-46fa-ad51-6b6b2a4581f6

Showing the scripts and css in use with an example of datatables

Comment (cabaceira) [reply apps/meteor/app/livechat/server/livechat.ts]: Corrected the issue with multiple-times loading the script. Please check it again.

File: apps/meteor/app/livechat/server/livechat.ts [hunk 5]
```diff
@@ -1,11 +1,37 @@
...
+const body = doc.querySelector('body');
+
+// TODO IMPROVE CONDITIONS CHECK
+const liveChatAdditionalScripts = settings.get<string>('Livechat_AdditionalWidgetScripts');
+if (liveChatAdditionalScripts) {
+	liveChatAdditionalScripts.split(',').forEach((script) => {
+		const scriptElement = doc.createElement('script');
+		scriptElement.src = script;
+		head?.appendChild(scriptElement);
+		body?.appendChild(scriptElement);
```
Comment (KevLehman) [inline apps/meteor/app/livechat/server/livechat.ts:22]: Do we need to put the same script in both?

File: apps/meteor/packages/rocketchat-i18n/i18n/en.i18n.json [hunk 6]
```diff
@@ -3088,6 +3088,8 @@
   "List_of_Direct_Messages": "List of Direct Messages",
   "List_view": "List View",
   "Livechat": "Livechat",
+  "Livechat_AdditionalWidgetScripts": "Livechat Widget Additional Scripts. Its possible to specify a CSV list of scripts, for example : https://yourUrl/customScrip1.js,https://yourUrl/customScrip2.js",
+  "Livechat_WidgetLayoutClasses": "Additional classes to add in the body section of the livechat widget. Its possible to specify a CSV list of classes, for example : https://yourUrl/customFile1.css,https://yourUrl/customFile2.css",
```
Comment (KevLehman) [inline apps/meteor/packages/rocketchat-i18n/i18n/en.i18n.json:3092]: Keep the name of the settings short and concise. Use the `_Description` of the setting to put examples and more data.

File: apps/meteor/ee/app/livechat-enterprise/server/settings.ts [hunk 7]
```diff
@@ -157,6 +157,32 @@ export const createSettings = async (): Promise<void> => {
 		});
 	});
 
+	await settingsRegistry.add('Livechat_AdditionalWidgetScripts', '', {
```
Comment (aleksandernsilva) [inline apps/meteor/ee/app/livechat-enterprise/server/settings.ts:160]: What naming convention is being used for these new settings? Shouldn't it be `Livechat_additional_widget_scripts`?

**Review by aleksandernsilva — CHANGES_REQUESTED** [review]

File: apps/meteor/packages/rocketchat-i18n/i18n/en.i18n.json [hunk 8]
```diff
@@ -3088,6 +3088,8 @@
   "List_of_Direct_Messages": "List of Direct Messages",
   "List_view": "List View",
   "Livechat": "Livechat",
+  "Livechat_AdditionalWidgetScripts": "Livechat Widget Additional Scripts. Its possible to specify a CSV list of scripts, for example : https://yourUrl/customScrip1.js,https://yourUrl/customScrip2.js",
```
Comment (KevLehman) [inline apps/meteor/packages/rocketchat-i18n/i18n/en.i18n.json:3091]: Separate this into 2:
```suggestion
  "Livechat_WidgetLayoutClasses": "Livechat Widget Additional Scripts",
  "Livechat_WidgetLayoutClasses_Description": "A comma-separated list of scripts, for example: https://yourUrl/customScrip1.js,https://yourUrl/customScrip2.js"

```

Notice how we use the same setting name followed by `_Description`, this is an internal convention and the code will use it automatically, without having you to specify the `i18nDescription` key.

Comment (MartinSchoeler) [inline apps/meteor/packages/rocketchat-i18n/i18n/en.i18n.json:3091]:
```suggestion
"Livechat_additional_widget_scripts": "Livechat Widget Additional Scripts"
"Livechat_additional_widget_scripts_Description": "Its possible to specify a CSV list of scripts, for example : https://yourUrl/customScrip1.js,https://yourUrl/customScrip2.js"
```

Comment (KevLehman) [reply apps/meteor/packages/rocketchat-i18n/i18n/en.i18n.json:3091]: Same change should be done to the other setting

Comment (MartinSchoeler) [reply apps/meteor/packages/rocketchat-i18n/i18n/en.i18n.json:3091]: We should use a pair of title & descriptions on the settings, this suggestion applies to the classes as well

File: apps/meteor/ee/app/livechat-enterprise/server/settings.ts (hunk 1 above)
Comment (MartinSchoeler) [inline apps/meteor/ee/app/livechat-enterprise/server/settings.ts:168]:
```suggestion
		i18nLabel: 'Livechat_additional_widget_scripts',
		i18nDescription: 'Livechat_additional_widget_scripts_Description',
```

This way you will be able to display the description as a subtext on the field, like this

Screenshot 2023-12-20 at 14 53 11

The same should be done for the classes setting

**Review by MartinSchoeler — CHANGES_REQUESTED** [review]

**Review by KevLehman — CHANGES_REQUESTED** [review]

File: apps/meteor/app/livechat/server/livechat.ts (hunk 5 above)
Comment (cabaceira) [reply apps/meteor/app/livechat/server/livechat.ts:22]: There may be edge cases where there is no head on the DOM where we should load the script in the body element, but i agree that we should not load the script twice. I will proceed with only appending it to in the head of the document.

File: apps/meteor/ee/app/livechat-enterprise/server/settings.ts [hunk 9]
```diff
@@ -157,6 +157,32 @@ export const createSettings = async (): Promise<void> => {
...
+		enableQuery: [omnichannelEnabledQuery],
+		modules: ['livechat-enterprise'],
+	});
+
+	await settingsRegistry.add('Livechat_WidgetLayoutClasses', '', {
+		type: 'string',
+		group: 'Omnichannel',
+		section: 'Livechat',
+		enterprise: true,
+		invalidValue: "",
```
Comment (KevLehman) [inline apps/meteor/ee/app/livechat-enterprise/server/settings.ts:178]:
```suggestion
		invalidValue: '',
```

File: apps/meteor/ee/app/livechat-enterprise/server/settings.ts [hunk 10]
```diff
@@ -157,6 +157,32 @@ export const createSettings = async (): Promise<void> => {
 		});
 	});
 
+	await settingsRegistry.add('Livechat_additional_widget_scripts', '', {
+		type: 'string',
+		group: 'Omnichannel',
+		section: 'Livechat',
+		enterprise: true,
+		invalidValue: "",
```
Comment (KevLehman) [inline apps/meteor/ee/app/livechat-enterprise/server/settings.ts:165]:
```suggestion
		invalidValue: '',
```

File: apps/meteor/packages/rocketchat-i18n/i18n/en.i18n.json [hunk 11]
```diff
@@ -3094,6 +3094,10 @@
   "List_of_Direct_Messages": "List of Direct Messages",
   "List_view": "List View",
   "Livechat": "Livechat",
+  "Livechat_additional_widget_scripts": "Livechat Widget Additional Scripts",
+  "Livechat_additional_widget_scripts_Description": "Its possible to specify a CSV list of scripts, for example : https://yourUrl/customScrip1.js,https://yourUrl/customScrip2.js",
```
Comment (KevLehman) [inline apps/meteor/packages/rocketchat-i18n/i18n/en.i18n.json:3098]:
```suggestion
  "Livechat_additional_widget_scripts_Description": "Setting makes possible to embed additional JS scripts that will be part of the widget bundle. You can add them as a list separated by commas, for example: `https://yourUrl/customScrip1.js,https://yourUrl/customScrip2.js`",
```

**Comment by ggazzo** [comment]
Aren't we going the exact opposite way to custom scripts?

File: apps/meteor/app/livechat/server/livechat.ts [hunk 12]
```diff
@@ -1,11 +1,35 @@
...
+if (liveChatAdditionalScripts) {
+	liveChatAdditionalScripts.split(',').forEach((script) => {
+		const scriptElement = doc.createElement('script');
+		scriptElement.src = script;
+		body?.appendChild(scriptElement);
+	});
+}
+
+const additionalClass = settings.get<string>('Livechat_WidgetLayoutClasses');
+if (additionalClass) {
```
Comment (KevLehman) [inline apps/meteor/app/livechat/server/livechat.ts:25]: Remove the `ifs` so an empty value means we remove the data

File: apps/meteor/app/livechat/server/livechat.ts [hunk 13]
```diff
@@ -1,11 +1,35 @@
...
+	additionalClass.split(',').forEach((css) => {
+		const linkElement = doc.createElement('link');
+		linkElement.rel = 'stylesheet';
+		linkElement.href = css;
+		head?.appendChild(linkElement);
+	});
+}
```
[the remark this answers is not in the document]
Comment (KevLehman) [reply apps/meteor/app/livechat/server/livechat.ts:32]: You can cache the function with mem and give it a 60s cache time. That should be fair considering the setting is not that would be changed that often (i assume)

[the remark this answers is not in the document]
Comment (cabaceira) [reply apps/meteor/app/livechat/server/livechat.ts]: done according to suggestions

**Review by MartinSchoeler — DISMISSED** [review]

**Review by KevLehman — APPROVED** [review]

**Comment by ggazzo** [comment]
I am not seeing specific code for CORS. This could be a significant issue because we do not want to allow custom scripts from the widget to be inserted being inserted on the main application.

**Comment by ggazzo** [comment]
also I'm not seeing the changeset and the jira task

**Pull request merged**